src/V3Expand.cpp

```diff
@@ -39,38 +39,6 @@
 
 VL_DEFINE_DEBUG_FUNCTIONS;
 
-//######################################################################
-// Find nodes with side effects, to mark as non-expandable
-
-class ExpandOkVisitor final : public VNVisitor {
-private:
-    // NODE STATE
-    //  AstNode::user2()        -> bool.  Is pure (along with all children)
-    const VNUser2InUse m_inuser2;
-
-    // STATE - for current visit position (use VL_RESTORER)
-    // Tracks similar to AstNode::isTreePureRecurse(), but avoid O(n^2)
-    // False = pure, as nodes that ExpandVisitor inserts preserve pureness
-    bool m_isImpure = true;  // Currently pure
-
-    void visit(AstNode* nodep) override {
-        bool selfImpure = !nodep->isPure();
-        {
-            VL_RESTORER(m_isImpure);
-            m_isImpure = false;
-            iterateChildren(nodep);
-            selfImpure |= m_isImpure;
-            nodep->user2(selfImpure);
-        }
-        m_isImpure |= selfImpure;
-    }
```

Comment:
The old code would disable this if there was any impure code anywhere under what was to be optimized.  I don't think this is the same as looking at pure() for the specific node being optimized, so I think you'll now optimize things that wouldn't have been before - and due to the cloning here that might result in incorrect expansions.  So I think we want to keep ExpandOkVisitor, checking isPure() similar to previously.

Reply:
I think you are right. As I briefly checked, the only nodes in which it may happen are AstNodeAssign. Their `pure` methods return `true`, even if they have impure LHS/RHS. I will restore the visitor.